### PR TITLE
Upgrade to terraform sdk v2

### PR DESCRIPTION
**Pull request by skarimo**
This PR updates the terraform provider to use terraform SDK v2. Closes https://github.com/DataDog/terraform-provider-datadog/issues/875

### CODE REVIEW

File: datadog/resource_datadog_timeboard.go [hunk 1]
```diff
@@ -1,6 +1,7 @@
 package datadog
```
Comment (zippolyte) [inline datadog/resource_datadog_timeboard.go:1]: delete that entirely

File: datadog/tests/data_source_datadog_dashboard_list_test.go [hunk 2]
```diff
@@ -16,20 +16,19 @@ func TestAccDatadogDashboardListDatasource(t *testing.T) {
...
-		PreCheck:     func() { testAccPreCheck(t) },
-		Providers:    accProviders,
-		CheckDestroy: testAccCheckDatadogDashListDestroy(accProvider),
+		PreCheck:          func() { testAccPreCheck(t) },
+		ProviderFactories: accProviders,
+		CheckDestroy:      testAccCheckDatadogDashListDestroy(accProvider),
 		Steps: []resource.TestStep{
 			{
-				Config:             testAccDatasourceDashboardListNameFilterConfig(uniq),
-				ExpectNonEmptyPlan: true,
```
Comment (zippolyte) [inline datadog/tests/data_source_datadog_dashboard_list_test.go:25]: is this not applicable anymore ?

File: datadog/tests/data_source_datadog_monitor_test.go [hunk 3]
```diff
@@ -66,33 +57,31 @@ func checkDatasourceAttrs(accProvider *schema.Provider, uniq string) resource.Te
...
 		resource.TestCheckResourceAttr(
 			"data.datadog_monitor.foo", "monitor_thresholds.0.critical_recovery", "0.7"),
 		resource.TestCheckResourceAttr(
 			"data.datadog_monitor.foo", "require_full_window", "true"),
 		resource.TestCheckResourceAttr(
 			"data.datadog_monitor.foo", "locked", "false"),
-		// Tags are a TypeSet => use a weird way to access members by their hash
-		// TF TypeSet is internally represented as a map that maps computed hashes
-		// to actual values. Since the hashes are always the same for one value,
-		// this is the way to get them.
```
Comment (zippolyte) [inline datadog/tests/data_source_datadog_monitor_test.go:91]: Are typeSet nice now with v2 ?

File: datadog/tests/resource_datadog_timeboard_test.go [hunk 4]
```diff
@@ -10,9 +10,9 @@ import (
 	"github.com/terraform-providers/terraform-provider-datadog/datadog"
```
Comment (zippolyte) [inline datadog/tests/resource_datadog_timeboard_test.go:10]: let's remove

File: datadog/tests/resource_datadog_screenboard_test.go [hunk 5]
```diff
@@ -9,9 +9,9 @@ import (
```
Comment (zippolyte) [inline datadog/tests/resource_datadog_screenboard_test.go:9]: let's remove

**Review by zippolyte — CHANGES_REQUESTED** [review]
A few comments. Once they are addressed, I think we can merge to the v2 branch, and iterate from here.

Also, do cassette really need rerecording ? Seems to me every change is internal, and shouldn't impact the requests made, isn't it ?

File: datadog/data_source_datadog_monitor.go [hunk 6]
```diff
@@ -63,45 +65,10 @@ func dataSourceDatadogMonitor() *schema.Resource {
...
-							Computed: true,
-						},
-					},
-				},
-			},
 			"monitor_thresholds": {
 				Description: "Alert thresholds of the monitor.",
 				Type:        schema.TypeList,
 				Computed:    true,
-				MaxItems:    1,
```
[the remark this answers is not in the document]
Comment (skarimo) [reply datadog/data_source_datadog_monitor.go:104]: tf plugin sdk V2 has more robust validations: https://www.terraform.io/docs/extend/guides/v2-upgrade-guide.html#stronger-validation-for-helper-schema-schema-computed-fields

Setting `MaxItems` on `Computed` fields is no longer supported and will throw an error

[the remark this answers is not in the document]
Comment (zippolyte) [reply datadog/data_source_datadog_monitor.go:104]: Ah true, didn't realise it was a datasource

File: datadog/provider.go [hunk 7]
```diff
@@ -45,86 +44,88 @@ func init() {
...
 // Provider returns the built datadog provider object
-func Provider() terraform.ResourceProvider {
-	utils.DatadogProvider = &schema.Provider{
-		Schema: map[string]*schema.Schema{
-			"api_key": {
-				Type:        schema.TypeString,
-				Optional:    true,
-				DefaultFunc: schema.MultiEnvDefaultFunc([]string{"DATADOG_API_KEY", "DD_API_KEY"}, nil),
-				Description: "(Required unless validate is false) Datadog API key. This can also be set via the DD_API_KEY environment variable.",
+func Provider() func() *schema.Provider {
```
[the remark this answers is not in the document]
Comment (skarimo) [reply datadog/provider.go:47]: Good catch thanks!

File: datadog/resource_datadog_dashboard.go [hunk 8]
```diff
@@ -117,48 +118,35 @@ func resourceDatadogDashboard() *schema.Resource {
...
-				return resource.RetryableError(fmt.Errorf("dashboard not created yet"))
-			}
-			return resource.NonRetryableError(err)
-		}
-
-		// We only log the error, as failing to update the list shouldn't fail dashboard creation
-		updateDashboardLists(d, providerConf, *dashboard.Id)
-
-		return resource.NonRetryableError(updateDashboardState(d, &getDashboard))
-	})
```
[the remark this answers is not in the document]
Comment (skarimo) [reply datadog/resource_datadog_dashboard.go:147]: Any reason this was added here? My thought process was we receive the created dashboard from the Create api call and can use that to set state, reducing number of API calls: https://github.com/DataDog/terraform-provider-datadog/blob/master/datadog/resource_datadog_dashboard.go#L128

File: datadog/resource_datadog_monitor.go [hunk 9]
```diff
@@ -555,23 +476,24 @@ func getUnmutedScopes(d *schema.ResourceData) []string {
...
 }
 
-func resourceDatadogMonitorCreate(d *schema.ResourceData, meta interface{}) error {
+func resourceDatadogMonitorCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
 	providerConf := meta.(*ProviderConfiguration)
 	datadogClientV1 := providerConf.DatadogClientV1
 	authV1 := providerConf.AuthV1
 
 	m, _ := buildMonitorStruct(d)
+	time.Sleep(3 * time.Second)
```
[the remark this answers is not in the document]
Comment (skarimo) [reply datadog/resource_datadog_monitor.go:485]: Yes thanks!

File: datadog/resource_datadog_service_level_objective.go [hunk 10]
```diff
@@ -190,7 +183,7 @@ func resourceDatadogServiceLevelObjectiveCustomizeDiff(diff *schema.ResourceDiff
 		for _, v := range attr.(*schema.Set).List() {
 			// Check that each monitor being added to the SLO exists
 			if _, _, err := datadogClientV1.MonitorsApi.GetMonitor(authV1, int64(v.(int))).Execute(); err != nil {
-				return utils.TranslateClientError(err, "error finding monitor to add to SLO")
+				return fmt.Errorf("error finding monitor to add to SLO: %s", err.Error())
```
[the remark this answers is not in the document]
Comment (skarimo) [reply datadog/resource_datadog_service_level_objective.go:186]: debug remnant as well. Thanks

File: datadog/tests/data_source_datadog_monitor_test.go (hunk 3 above)
Comment (zippolyte) [reply datadog/tests/data_source_datadog_monitor_test.go:91]: sweet

**Review by zippolyte — APPROVED** [review]
Looks good, I think we should merge this, and keep working on that branch to polish things if needed

**Pull request merged**